devito/function.py

```diff
@@ -355,7 +356,12 @@ def argument_values(self, alias=None, **kwargs):
             else:
                 # We've been provided a pure-data replacement (array)
                 values[key] = new
-                # TODO: Re-derive defaults from shape of array
+                if len(new.shape) != len(self.indices):
```

Comment:
Here's my last concern. Shouldn't this whole block ``if-then-else`` be promoted one level up? ie shouldn't it be nested under ``if self.name in kwargs``, rathen than under the first ``else``?

the check on the length of the shape should per performed always, correct? btw, instead of ``len(self.indices)``, you should use ``self.ndim`` (defined in ``types.py``)

Reply:
True, should be checked one level up